mesonbuild/backend/xcodebackend.py

```diff
@@ -396,10 +405,23 @@ def get_xcodetype(self, fname: str) -> str:
 
     def generate_filemap(self) -> None:
         self.filemap = {} # Key is source file relative to src root.
+        self.foldermap = {}
         self.target_filemap = {}
         for name, t in self.build_targets.items():
             for s in t.sources:
                 if isinstance(s, mesonlib.File):
+                    if '/' in s.fname:
+                        # From the top level down, add the folders containing the source file.
+                        folder = s.fname.split('/')[:-1]
```

Comment:
I think it would make it more clear we're dealing with paths if you did:
```python
folder = os.path.split(os.path.dirname(f.fname))
```

And below did:
```python
fpath = os.path.join(*folder)
```